File: API/src/main/java/tororo1066/uniqueitemprovider/IUniqueItemProvider.java
package tororo1066.uniqueitemprovider;

import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface IUniqueItemProvider {

    IUniqueItem getUniqueItem(@NotNull ItemStack itemStack);

    @NotNull IUniqueItem getOrCreateUniqueItem(@NotNull ItemStack itemStack);

    @NotNull List<AbstractItemProvider> getProviders(@NotNull ItemStack itemStack);

    void registerProvider(@NotNull AbstractItemProvider provider);
}

File: API/src/main/kotlin/tororo1066/uniqueitemprovider/IUniqueItemProvider.kt
package tororo1066.uniqueitemprovider

import org.bukkit.inventory.ItemStack

interface IUniqueItemProvider {

    fun getUniqueItem(itemStack: ItemStack): IUniqueItem?

    fun getOrCreateUniqueItem(itemStack: ItemStack): IUniqueItem

    fun getProviders(itemStack: ItemStack): List<AbstractItemProvider>

    fun registerProvider(provider: AbstractItemProvider)
}